Lib/unittest/mock.py

```diff
@@ -423,6 +423,36 @@ def __init__(
             parent=None, _spec_state=None, _new_name='', _new_parent=None,
             _spec_as_instance=False, _eat_self=None, unsafe=False, **kwargs
         ):
+        if parent is not None and not isinstance(parent, NonCallableMock):
+            import warnings
+            if isinstance(self, NonCallableMagicMock):
+                stacklevel = 3
+            elif isinstance(self, MagicMock):
+                stacklevel = 4
+            elif isinstance(self, Mock):
+                stacklevel = 3
+            elif isinstance(self, NonCallableMock):
+                stacklevel = 2
+            else:
+                stacklevel = 1
+            warnings.warn('parent argument should be a Mock instance',
+                          DeprecationWarning, stacklevel=stacklevel)
+
+        if name is not None and not isinstance(name, str):
+            import warnings
+            if isinstance(self, NonCallableMagicMock):
+                stacklevel = 3
+            elif isinstance(self, MagicMock):
+                stacklevel = 4
+            elif isinstance(self, Mock):
+                stacklevel = 3
+            elif isinstance(self, NonCallableMock):
+                stacklevel = 2
+            else:
+                stacklevel = 1
+            warnings.warn('name argument should be a string',
+                          DeprecationWarning, stacklevel=stacklevel)
```

Comment:
This would be a backwards incompatible change with exceptions. Right now, using non-string arg is fine if you don't display the mock obj.

That said, I think this probably shouldn't be merged because it doesn't seem like people are running into this issue, as it was created a while ago and nobody else commented on it.

If you agree, we can close this issue and MR.